sentinel/src/lib/purging.js

```diff
@@ -348,23 +356,102 @@ const batchedContactsPurge = (roles, purgeFn, startKey = '', startKeyDocId = '')
+  const getQueryParams = (startKeyDocId) => ({
     limit: BATCH_SIZE,
     key: JSON.stringify('_unassigned'),
     startkey_docid: startKeyDocId,
     include_docs: true
+  });
+
+  const purgeCallback = (rolesHashes, rows) => {
+    const toPurge = {};
+    rows.forEach(row => {
+      const doc = row.doc;
+      rolesHashes.forEach(hash => {
+        toPurge[hash] = toPurge[hash] || {};
+        const purgeIds = doc.form ?
+          purgeFn({ roles: roles[hash] }, {}, [doc], []) :
+          purgeFn({ roles: roles[hash] }, {}, [], [doc]);
+
+        if (!validPurgeResults(purgeIds)) {
+          return;
+        }
+        toPurge[hash][doc._id] = purgeIds.includes(doc._id);
+      });
+    });
+
+    return toPurge;
   };
 
+  return batchedPurge(type, url, getQueryParams, purgeCallback, roles, '');
+};
+
+const batchedTasksPurge = (roles) => {
+  const type = 'tasks';
+  const url = `${db.couchUrl}/_design/medic/_view/tasks_by_state`;
+
+  const terminalStates = ['Cancelled', 'Completed', 'Failed'];
+  const getQueryParams = (startKeyDocId) => ({
+    limit: BATCH_SIZE,
+    keys: JSON.stringify(terminalStates),
+    startkey_docid: startKeyDocId,
+  });
+
+  const purgeCallback = (rolesHashes, rows) => {
+    const now = moment();
+    const toPurge = {};
+    rows.forEach(row => {
+      rolesHashes.forEach(hash => {
+        toPurge[hash] = toPurge[hash] || {};
+        const daysSinceTaskEndDate = now.diff(row.value.endDate, 'days');
+        if (daysSinceTaskEndDate >= TASK_EXPIRATION_PERIOD) {
+          toPurge[hash][row.id] = row.id;
+        }
+      });
+    });
+    return toPurge;
+  };
+
+  return batchedPurge(type, url, getQueryParams, purgeCallback, roles, '');
+};
+
+const batchedTargetsPurge = (roles) => {
+  const type = 'targets';
+  const url = `${db.couchUrl}/_all_docs`;
+
+  const lastAllowedReportingIntervalTag = moment().subtract(TARGET_EXPIRATION_PERIOD, 'months').format('YYYY-MM');
+  const getQueryParams = (startKeyDocId) => ({
+    limit: BATCH_SIZE,
+    start_key: JSON.stringify(startKeyDocId),
+    end_key: JSON.stringify(`target~${lastAllowedReportingIntervalTag}~`),
+  });
+
+  const purgeCallback = (rolesHashes, rows) => {
+    const toPurge = {};
+    rows.forEach(row => {
+      rolesHashes.forEach(hash => {
+        toPurge[hash] = toPurge[hash] || {};
+        toPurge[hash][row.id] = row.id;
+      });
+    });
+    return toPurge;
+  };
+
+  return batchedPurge(type, url, getQueryParams, purgeCallback, roles, 'target~');
+};
+
+const batchedPurge = (type, uri, getQueryParams, purgeCallback, roles, startKeyDocId) => {
+  let nextKeyDocId;
+  const rows = [];
+  const docIds = [];
+  const rolesHashes = Object.keys(roles);
+
+  logger.debug(`Starting ${type} purge batch with id ${startKeyDocId}`);
   return request
-    .get(`${db.couchUrl}/_design/medic/_view/docs_by_replication_key`, { qs: queryString, json: true })
+    .get(uri, { qs: getQueryParams(startKeyDocId), json: true })
```

Comment:
Why are we not using Pouch here? Obviously it'd need a little massaging for `allDocs` vs `query` but it'd save us a lot of URL and parameter manipulation.

Reply:
Because PouchDB doesn't support `start_key_doc_id` :(

Reply:
Ugh, I remember now, sorry!